### PR TITLE
Code Modularization and Testing

### PR DESCRIPTION
**Pull request by nitinramadoss**
- Adds bash script that downloads new survey responses through Survey API
- Adds survey responses after running the script
- Makes code more modular to enable unit testing, primarily data store testing through the file parsing classes

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/data/ReadData.java [hunk 1]
```diff
@@ -1,66 +1,84 @@
...
-import java.util.Set;
 
 public class ReadData {
-  private DatastoreService datastoreService;
   private SentimentData sentimentService;
-  private ArrayList<Entity> newEntities;
   private final String DELIMITER = ",";
 
+  /**
+   * Default constructor with no sentiment analysis dependency. Objects initialized through this
```
Comment (prashantiA) [inline capstone/src/main/java/com/google/sps/data/ReadData.java:19]: I really appreciate your use of documentation here!

File: capstone/src/main/java/com/google/sps/data/ReadData.java [hunk 2]
```diff
@@ -85,75 +103,63 @@ public void readFile(FileReader file, String zipCode) {
...
 
-        if (text.length() > 0) {
+        if (sentimentService != null && text.length() > 0) {
           score = sentimentService.getSentiment(text);
         }
 
-        long responseTimeOne = Long.parseLong(values[indexOfLong++]);
-        long responseTimeTwo = Long.parseLong(values[indexOfLong++]);
-        long responseTimeThree = Long.parseLong(values[indexOfLong]);
+        int responseTimeOne = (int) Math.round(Double.parseDouble(values[indexOfInt++]));
```
Comment (prashantiA) [inline capstone/src/main/java/com/google/sps/data/ReadData.java:118]: Why did you end up using ints here instead of doubles?

File: capstone/src/main/java/com/google/sps/data/ReadData.java [hunk 3]
```diff
@@ -85,75 +103,63 @@ public void readFile(FileReader file, String zipCode) {
...
-        Entity inStore; // Entity that may or may not exist in Datastore
 
-        try {
-          inStore = datastoreService.get(entity.getKey());
-        } catch (EntityNotFoundException e) {
-          inStore = null;
+        entity.setProperty("zipCode", zipCode);
+        entity.setProperty("id", id);
+        entity.setProperty("date", date);
+        entity.setProperty("completionStatus", completionStatus);
```
Comment (prashantiA) [inline capstone/src/main/java/com/google/sps/data/ReadData.java:127]: In the future I think it would be really cool if we could create a file just with all the property name constants in it and then all draw from there to help make our code consistent and guard against typos 😄

File: capstone/src/main/java/com/google/sps/data/ReadData.java [hunk 4]
```diff
@@ -85,75 +103,63 @@ public void readFile(FileReader file, String zipCode) {
...
-          }
-
-          if (total != 0) {
-            entity.setProperty("averageHouseholdIncome", Math.round(sumIncome / total));
-            entity.setProperty("crimeRate", Math.round(sumCrimeRate / total));
-            entity.setProperty("policeStationRating", Math.round(sumStationRating / total));
-          }
-
-          newEntities.add(entity);
+        if (total != 0) {
```
Comment (prashantiA) [inline capstone/src/main/java/com/google/sps/data/ReadData.java:155]: If total is 0 do we want to set the rating/crime rate/average income to something to indicate that there's not data?

File: capstone/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -12,8 +18,29 @@
 public class DataServlet extends HttpServlet {
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ReadData readData = new ReadData();
-    readData.readAll();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (prashantiA) [inline capstone/src/main/java/com/google/sps/servlets/DataServlet.java:21]: Can you add a comment explaining what this servlet does?

File: capstone/src/main/webapp/surveys/survey-download.sh [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  echo Store the name of the json file where the client secret is stored within the CLIENT_SECRET_FILE variable 
+  exit 1
+fi
+
+#download oauth21 tool
+wget https://storage.googleapis.com/oauth2l/latest/linux_amd64.tgz
+tar -xvf linux_amd64.tgz
+
+#id for each survey that is currently deployed
+SURVEY_IDS=('iilszvt24alys4gp7yyhypsuuq' 'edxr3yho5rkjsbcrepg7x4373e' 'wfj4mvsfw3mlcjvkoushioiomm' 'el6cjg2mb6rjbo5sel7lm3kkj4' 'rvhtbkditxingqxn5qoixau2eu'
```
Comment (prashantiA) [inline capstone/src/main/webapp/surveys/survey-download.sh:17]: A future improvement might be to store these in a separate file instead of in the bash script.

File: capstone/src/main/webapp/surveys/survey-download.sh [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+'cm3hdousx3jvt4uaf4o6ukkjrq' 'lla7pgdxkjkrnxdjxnzzt24y7y' 'xsz7qvuip42ibjtsltglv7cxsy' 'lhujzdwrs6ukp4ci5rz734wy7y' 'h77xuhlpouhhlnc52lh7r6bflm' 
+'qqegqs3pzo7aa7unu6f3wskeiq' '7m777v2tsvldrcjapn5fqiz3xy' '2qe3qoubki66ixvvr2brglahwi' 'oef7cqircmjizw6uk4r4xaukdi' 'rb2v6nh6vk4keyfblobuxfve2m' 
+'mywyxzaekbuh43kc76fe47lxke' 'mg7qdviq7idbi6wia6t7svqgfm' 'dglzpokdznwwb7mbv27idkewbm' 'jdq4devbyy2kuup2minggrs5ty' 'mzulceapcv5nm3q3jcewfv3l2y' 
+'z4gkxlmetp3cgahkskuuwno2tu' 'hjgt4ozk7uztw2eylsjrslls5u' 'hydgtwhoo63meppzhsicdtimhq' 'qqbzaiu2ptb5xrpaoipbz4eosi' 'opjq7l3y267o2rdljpvcs37qqa' 
+'pcbt55lli3ypmo4p26veqvh5sm' 'u5s5nbi7fwdzycztqqjmpndseq' '7waxhohyj4aty2kxiu6lwq6zj4' 'ufgsknnuyr2sbilluju6hd4ycu' 'oc4ehskpfnbpde5hhgp3gtcabq' 
+'v6zee7g66cg74klpr6nzo4vvjq' 'jspnt7lj5c4i2kv4fcudbml23q' 'btsrnwdc62mcavmfefql7cvml4' 'vsxcry3luael7zz7y6saf5e7ta' '65zghd6jjc2ffnyszgsgnbuadm' 
+'2wes62yvhtknqkwbmkd5dot47a' 'qhbxjs7nd5a4zxnyfyudsogmva')
+
+ZIP_CODES=('94103-r' '94103-r2' '94107-r' '94107-r2' '94109-r' '94109-r2' '94110-r' '94110-r2' '94112-r' '94112-r2' '94114-r' '94114-r2' '94116-r' '94116-r2' '94117-r' '94117-r2' '94118-r'
```
Comment (prashantiA) [inline capstone/src/main/webapp/surveys/survey-download.sh:26]: Do these correspond to the survey IDs above? A comment would be really useful here 😄

File: capstone/src/test/java/com/google/sps/data/DataStoreTest.java [hunk 8]
```diff
@@ -34,14 +41,99 @@ private void doTest() {
...
+    Entity b = getMockEntity("ResponseTest", "94151", 2, 0.83f);
+
+    testEntities.add(a);
+    testEntities.add(b);
+
+    datastore.put(testEntities);
+
+    Entity entity = new Entity("ResponseTest", 2);
+
+    assertTrue(inLocalStore(entity));
```
Comment (prashantiA) [inline capstone/src/test/java/com/google/sps/data/DataStoreTest.java:94]: For all of the asserts in this file -- Junit has an option that allows you to also add a string as an error message if the asserts fail, which can be useful for debugging if tests do not pass.

File: capstone/src/test/java/com/google/sps/data/DataStoreTest.java [hunk 9]
```diff
@@ -34,14 +41,99 @@ private void doTest() {
...
+  /**
+   * Creates an entity based on given survey response values
+   *
+   * @param kind Namespace in datastore
+   * @param zipCode Location of where survey was taken
+   * @param id Unique id given to each survey
+   * @param score Sentiment score for the survey response
+   * @return Entity data store object representing a survey response
+   */
+  public Entity getMockEntity(String kind, String zipCode, int id, float score) {
```
Comment (prashantiA) [inline capstone/src/test/java/com/google/sps/data/DataStoreTest.java:73]: I really like how you created a helper function for this! I ended up doing the same thing after seeing your code 😄

File: capstone/src/test/java/com/google/sps/data/DataStoreTest.java [hunk 10]
```diff
@@ -34,14 +41,99 @@ private void doTest() {
...
+   * @param file The file that will be parsed to create entities
+   * @return void
+   */
+  public void getEntities(ArrayList<Entity> entities, File file) throws IOException {
+    ReadData readData = new ReadData();
+    readData.entitiesFromFile(entities, file);
+  }
+
+  @Test
+  public void testInsertEntity() throws IOException {
```
Comment (prashantiA) [inline capstone/src/test/java/com/google/sps/data/DataStoreTest.java:110]: Seems like you just have a few normal csvs (which seems like typical data). What are some edge cases that you can think of? Some of the ones that come to mind are empty files and adding the same file twice

File: capstone/src/main/java/com/google/sps/data/ReadData.java (hunk 2 above)
Comment (nitinramadoss) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:118]: Doubles and floats were giving precision errors when stored in data store, so I stored them as ints. For example, if i wanted to store 217.48 I would store it as 21748 and then divide it by 100 when fetching.

File: capstone/src/main/java/com/google/sps/data/ReadData.java (hunk 3 above)
Comment (nitinramadoss) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:127]: I agree, having this many properties lead to a lot of issues when refactoring property names

File: capstone/src/main/java/com/google/sps/data/ReadData.java (hunk 4 above)
Comment (nitinramadoss) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:155]: I believe it's set to null by default in data store

File: capstone/src/main/webapp/surveys/survey-download.sh (hunk 7 above)
Comment (nitinramadoss) [reply capstone/src/main/webapp/surveys/survey-download.sh:26]: I'll add a comment explaining it, but yes they match the index of the survey id

File: capstone/src/test/java/com/google/sps/data/DataStoreTest.java (hunk 10 above)
Comment (nitinramadoss) [reply capstone/src/test/java/com/google/sps/data/DataStoreTest.java:110]: Fortunately, the surveys api ensures that the survey responses are complete when downloaded through the script. Duplicate files are written as one file by the script as well

File: capstone/src/main/java/com/google/sps/data/ReadData.java (hunk 2 above)
Comment (lee12jin) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:118]: +1 doubles and floats give precision errors. This is an accepted practice but just to be clear, you may want to rename the variable to `responseTimeOneCentis` if that's easy to do. Either that or leave documentation indicating that we store in the hundreds.

File: capstone/src/main/java/com/google/sps/data/ReadData.java (hunk 4 above)
Comment (lee12jin) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:155]: Would we want to set those to 0 or null?

File: capstone/src/main/webapp/surveys/survey-download.sh [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+linux_amd64/oauth2l fetch --credentials $CLIENT_SECRET_FILE --scope https://www.googleapis.com/auth/surveys
+
+for survey in "${SURVEY_IDS[@]}"; do 
+  TOKEN=$(linux_amd64/oauth2l fetch --credentials $CLIENT_SECRET_FILE --scope https://www.googleapis.com/auth/surveys)
+  curl "https://surveys.google.com/reporting/export_oauth?format=xls&survey=$survey&access_token=$TOKEN" > $survey.xls
+done
+echo Finished downloading ${#SURVEY_IDS[@]} surveys
+
+#Download xls to csv module
+echo Downloading CSV converter
```
Comment (lee12jin) [inline capstone/src/main/webapp/surveys/survey-download.sh:40]: Wrap this into quotations

File: capstone/src/main/webapp/surveys/survey-download.sh [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+#If prompted, follow instructions and find verification code in url param code
+linux_amd64/oauth2l fetch --credentials $CLIENT_SECRET_FILE --scope https://www.googleapis.com/auth/surveys
+
+for survey in "${SURVEY_IDS[@]}"; do 
+  TOKEN=$(linux_amd64/oauth2l fetch --credentials $CLIENT_SECRET_FILE --scope https://www.googleapis.com/auth/surveys)
+  curl "https://surveys.google.com/reporting/export_oauth?format=xls&survey=$survey&access_token=$TOKEN" > $survey.xls
+done
+echo Finished downloading ${#SURVEY_IDS[@]} surveys
+
+#Download xls to csv module
```
Comment (lee12jin) [inline capstone/src/main/webapp/surveys/survey-download.sh:39]: This comment doesn't provide much more than what's already discussed below. I would leave it out.

File: capstone/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -12,8 +18,29 @@
...
+    ArrayList<Entity> allEntities = readData.allEntitiesFromFiles();
+
+    for (int i = 0; i < allEntities.size(); i++) {
+      Entity currentEntity = allEntities.get(i);
+      Entity inStore = null;
+
+      try {
+        inStore = datastore.get(currentEntity.getKey());
+      } catch (EntityNotFoundException e) {
+        inStore = null;
```
Comment (lee12jin) [inline capstone/src/main/java/com/google/sps/servlets/DataServlet.java:33]: You don't need to set inStore again here, you can just continue the for loop by adding a `continue` and the rest of this logic won't run.

Comment (lee12jin) [reply capstone/src/main/java/com/google/sps/servlets/DataServlet.java:33]: This logic including the logic below.

File: capstone/src/main/java/com/google/sps/data/ReadData.java [hunk 14]
```diff
@@ -1,66 +1,84 @@
...
-  public void readFile(FileReader file, String zipCode) {
-    final int EXPECTED_LENGTH = 15;
+  public void entitiesFromFile(ArrayList<Entity> newEntities, File file) {
+    String zipCode = file.getName().substring(0, 5);
+    FileReader fileReader = null;
+
+    try {
+      fileReader = new FileReader(file);
+    } catch (FileNotFoundException e) {
+      System.out.println("No file found");
```
Comment (jasperlee-google) [inline capstone/src/main/java/com/google/sps/data/ReadData.java:73]: Should we just return early here if file isn't found or raise the exception?

Comment (nitinramadoss) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:73]: I did not think of that, thank you for the suggestion!

File: capstone/src/test/java/com/google/sps/data/DataStoreTest.java (hunk 10 above)
Comment (prashantiA) [reply capstone/src/test/java/com/google/sps/data/DataStoreTest.java:110]: Its really cool that the surveys api does a lot of this for you. However, I think it would still be nice to show that your system is robust and can deal with these situations (which from glancing over your code it can).

File: capstone/src/main/java/com/google/sps/data/ReadData.java (hunk 4 above)
Comment (nitinramadoss) [reply capstone/src/main/java/com/google/sps/data/ReadData.java:155]: Either works in this case since I check if the property is null
